filebeat/channel/runner_mock_test.go

```diff
@@ -0,0 +1,118 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package channel
+
+import (
+	"testing"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/cfgfile"
+
+	conf "github.com/elastic/elastic-agent-libs/config"
+
+	"github.com/stretchr/testify/require"
+)
+
+type runnerFactoryMock struct {
+	cfgfile.RunnerFactory
```

Comment:
I disagree, this pattern does not force us to refactor all the mocks every time we change the interface. I don't see any problem with using it in this test because here we know for a fact what functions are supposed to be used and what are not.

Otherwise, I'd have to write a lot of unnecessary code and mostly empty functions just for nothing.

Reply:
To clarify, I'm not saying it forces us to refactor mocks when the interface changes, I'm saying it _prevents_ us from refactoring mocks even when it would be appropriate:

If `RunnerFactory` makes a small change to the parameters of `Create`, your code here would still satisfy the interface and build without problems, even though the implementation would no longer be called by anything.

Similarly, if a new method is added to `RunnerFactory`, and invoked during pipeline setup, this test would break with a nil reference even though there's no build-time indication that anything is missing.

Since this is a literal one-line change, I don't think it's really "A lot of unnecessary code just for nothing." I already spent hours this morning troubleshooting a nil pointer in the tests caused by this exact issue, since it prevents changed interfaces from showing up in static analysis.

Reply:
Changed in https://github.com/elastic/beats/pull/34870/commits/16c10e06c0498694bfd9bed007f0f154fa9c5ba8

I still disagree though.

> If RunnerFactory makes a small change to the parameters of Create, your code here would still satisfy the interface and build without problems, even though the implementation would no longer be called by anything.

In this case the test would fail, otherwise what's the point of the test.

> Similarly, if a new method is added to RunnerFactory, and invoked during pipeline setup, this test would break with a nil reference even though there's no build-time indication that anything is missing.

In my opinion, panic in tests is okay when you change the interface. If you run the tests (and we should do it all the time), you'll see the panic and where to fix it.

I've worked on huge code bases in Go before and it's very discouraging when you add a new function to an interface and dozens of tests cannot build anymore because every package has its own mock implementation that implements this interface.

I don't think it's better than fixing a panic on a rare occasion.

Reply:
Well -- for me, it is more discouraging to have a dozen tests that segfault in CI in other packages, than to have a dozen build failures where my IDE immediately tells me what piece is missing.

But I don't disagree with the pattern in all cases -- like I said, I think it makes sense here for `clientMock`, and it can also be reasonable for testing small pieces of a large interface. But if it can be avoided with a one-liner, I prefer to do so. Thank you for the changes.